pkg/network/tcp.go

```diff
@@ -51,30 +51,46 @@ func handleConnection(s *Server, conn net.Conn) {
 	peer := NewTCPPeer(conn, s)
 	s.register <- peer
 
-	// remove the peer from connected peers and cleanup the connection.
-	defer func() {
-		peer.disconnect()
-	}()
-
 	// Start a goroutine that will handle all outgoing messages.
 	go peer.writeLoop()
 	// Start a goroutine that will handle all incomming messages.
 	go handleMessage(s, peer)
+	// Start a goroutine that will handle polling for incoming message.
+	go pollPeerForMessage(s, conn, peer)
 
 	// Read from the connection and decode it into a Message ready for processing.
+	// Wait until
+	for {
+		select {
+		case <-peer.shouldDisconnect:
+			peer.disconnect()
+			break
+		}
+	}
+}
+
+// pollPeerForMessage polls for messages on the TCP connection and sends them
+// down p.receive for processing.
+func pollPeerForMessage(s *Server, conn net.Conn, p *TCPPeer) {
 	for {
 		msg := &Message{}
 		if err := msg.decode(conn); err != nil {
 			s.logger.Printf("decode error: %s", err)
 			break
 		}
 
-		peer.receive <- msg
+		if msg.Payload != nil {
```

Comment:
Not sure about that. Messages could have an empty payload. Take `verack` and `getaddr` for example, those are messages with no payload.

Reply:
@anthdm yea that was my initial thought on what the problem was but I then realised as you mentioned that some messages have an empty payload - that’s why it’s not in the PR - should have rebates the old commits out as not relevant to the PR 👍